DOCS-12973 DG semantic extensions (initial commit) (#58)
mulesoft/docs-anypoint-datagraph: modules/ROOT/pages/api-extensions.adoc

= Anypoint DataGraph API Extension Reference

> The following guide defines API extensions DataGraph. 

Reviewer: ```suggestion
This guide defines API extensions for DataGraph. 
```

== Object Type Extensions 

> This extension is used to mark that an object type has a primary key, and thus it can be used as reference or extension.

Reviewer: ```suggestion
This extension marks that an object type has a primary key, and thus it can be used as a reference or extension.
```

Reviewer: Global suggestion for lead-ins: 

Remove "is used to" as shown in the comment.

This extension applies to `ObjectType`.

=== Structure

[%header%autowidth.spread]
|===
|Field |Type |Description

> |`entity` |Map |Marks an object type as entity.

Reviewer: ```suggestion
|`entity` |Map |Marks an object type as an entity.
```

|`entity.key` |String |An array that must not be empty. Each item must reference a field that is a key of the entity.
|===

=== OAS Example

----
components:
  schemas:
    Customer:
  	x-entity:
    	  key: [id]
  	properties:
    	  name:
          type: string
    	  email:
          type: string
    	  id:
          type: string
    	  ...
----

=== RAML Example

----
types:
  Customer:
    (dg.entity):
      key: [id]
    properties:
    	name:
        type: string
    	email:
        type: string
    	id:
        type: string
    	...
----

> == Entity Reference

Reviewer 2: Our entity reference extension its used with "entity-references" instead of "entity-reference".
We had a typo in our document, and we will update it soon too.

Author: got it

> This extension is used to generate a DataGraph _link_ from the properties of a type to another type. The link can be to a type that is not defined in the same API spec.

Reviewer: Suggestion: "a type" > "one type"

This extension applies to `ObjectType`.

=== Structure

[%header%autowidth.spread]
|===
|Field |Type |Description

> |`entity-reference` |Array of objects |A list of all the references.

Reviewer: General global comment - descriptions should be consistent regarding whether they start with an article like "a" or "the" or not. They are inconsistent in most of the tables.

          fieldName: store
          keyMapping:
            storeId: id
      properties:
    	  orderItems:
          ...
    	  customerId:
          type: string
    	  storeId:
          type: string
    Customer:
  	x-entity:
    	  key: [id]
  	properties:
    	  name:
          type: string
    	  email:
          type: string
    	  id:
          type: string
    Store:
  	x-entity:
    	  key: [id]
  	properties:
    	  name:
          type: string
    	  id:
          type: string
    	  ...
----

=== RAML Example

----
types:
  Order:
    (dg.entity-reference):
      - name: Customer
        fieldName: customer
        keyMapping:
          customerId: id
      - name: Store
        fieldName: store
        keyMapping:
          storeId: id
    properties:
    	orderItems:
        ...
    	customerId:
        type: string
    	storeId:
        type: string
  Customer:
    (dg.entity):
      key: [id]
    properties:
    	name:
        type: string
    	email:
        type: string
    	id:
        type: string
    	...
  Store:
    (dg.entity):
      key: [id]
    properties:
    	name:
        type: string
    	id:
        type: string
----

== Entity Provider

> This extension is used to mark that an endpoint returns instances of an entity by id. In DataGraph this is the _default query method_.

Reviewer: ```suggestion
This extension is used to mark that an endpoint returns instances of an entity by id. In DataGraph, this is the _default query method_.
```

Reviewer: Suggest id ? ID

            	$ref: "#/components/schemas/Customer"
----

=== RAML Example

----
types:
  Customer:
    (dg.entity):
      key: [id]
    properties:
    	name:
        type: string
    	email:
        type: string
    	id:
        type: string
    	…

/customer/{customerId}:
  uriParameters:
    customerId: string
  get:
    (dg.entityProvider):
      keyMapping:
        customerId: id
    responses:
      200:
        application/json:
          body: Customer
----

== DataGraph-specific Extensions 

The main annotation `datagraph` cannot have the same name for different domains. Because of this, the annotations for operations, type, and object properties use postfixes:

* Operations: -method
* Type: -type
* Object properties: -field

=== Hide

Use this extension to mark an element as hidden. Hidden elements aren't shown in the unified schema, and DataGraph doesn't return them in any queries. 

This extension applies to:

* Operations
* Any type
* Object properties

=== Structure

The target is not specified because it is implicit in the operation.

[%header%autowidth.spread]
|===
|Field |Type |Description
| `datagraph` |Map |A container for DataGraph-specific annotations.

> | `datagraph.hide` |Boolean |If `true` the element is hidden in the unified schema. If `false` the element is not hidden, which provides the same result as not setting the annotation.

Reviewer: ```suggestion
| `datagraph.hide` |Boolean |If `true`, the element is hidden in the unified schema. If `false`, the element is not hidden, which provides the same result as not setting the annotation.
```

= Anypoint DataGraph API Extension Reference

The following guide defines API extensions DataGraph. 

There are two groups of extensions: Entity extensions and DataGraph extensions. You must represent each group as different extension instances and store them in Exchange as different assets.

> == Entity Extensions

Reviewer 2: Right now we decided to use a single group, datagraph extensions, instead of two.

These extensions define behaviors specific for object types. 

=== Entity

This extension is used to mark that an object type has a primary key, and thus it can be used as reference or extension.

This extensions applies to:

* ObjectType

==== Structure

[%header%autowidth.spread]
|===
|Field |Type |Description
|`entity` |Map |Marks an object type as entity.
|`entity.key` |String |Array, must not be empty. Each item must reference a field that is a key of the entity.
|===

==== Customization

Each annotation should result in a `PrimaryKeyCustomizations` set for each field referenced as key. 

==== OAS Example

----
components:
  schemas:
    Customer:
  	x-entity:
    	  key: [id]
  	properties:
    	  name:
          type: string
    	  email:
          type: string
    	  id:
          type: string
    	  ...
----

==== RAML Example

----
types:
  Customer:

>     (entity):

Author: ```suggestion
    (dg.entity):
```

>     (entity):

Author: @Extor3D I added the prefix throughout. Can you verify?

|`entity-reference.[n].fieldName` |String |Name of the new field that will be created for this type.
|`entity-reference.[n].keyMapping` |Map |The keys of the map represent fields of this type, and the values represents the keys of the referenced type.
|===

==== Customization

Each annotation should result in a `ReferenceCustomization`.  

==== OAS Example

----
components:
  Schemas:
    Order:
    	x-entity-reference:
        - name: Customer
          fieldName: customer
          keyMapping:
            customerId: id
        - name: Store
          fieldName: store
          keyMapping:
            storeId: id
      properties:
    	  orderItems:
          ...
    	  customerId:
          type: string
    	  storeId:
          type: string
    Customer:
  	x-entity:
    	  key: [id]
  	properties:
    	  name:
          type: string
    	  email:
          type: string
    	  id:
          type: string
    Store:
  	x-entity:
    	  key: [id]
  	properties:
    	  name:
          type: string
    	  id:
          type: string
    	  ...
----

==== RAML Example

----
types:
  Order:

>     (entity-reference):

Author: ```suggestion
    (dg.entity-reference):
```

      - name: Customer
        fieldName: customer
        keyMapping:
          customerId: id
      - name: Store
        fieldName: store
        keyMapping:
          storeId: id
    properties:
    	orderItems:
        ...
    	customerId:
        type: string
    	storeId:
        type: string
  Customer:

>     (entity):

Author: ```suggestion
    (dg.entity):
```

      key: [id]
    properties:
    	name:
        type: string
    	email:
        type: string
    	id:
        type: string
    	...
  Store:

>     (entity):

Author: ```suggestion
    (dg.entity):
```

      key: [id]
    properties:
    	name:
        type: string
    	id:
        type: string
----

=== Entity Provider

This extension is used to mark that an endpoint returns instances of an entity by id. In DataGraph this is the _default query method_.

This extension applies to:

* GET operations

==== Structure

The target is not specified because it is implicit in the operation.

[%header%autowidth.spread]
|===
|Field |Type |Description
|`entity-provider` |Map |Marks that this endpoint returns entity instances.
|`entity-reference.keyMapping` |Map |The keys of the map represent parameters of this operation, and the values represents the keys of the target type.
|===

==== Customization

> Each annotation should result in a `DefaultQueryMethodCustomization` and a set of ????	.

Author: @jisoldi do you know what should go here?

==== OAS Example

----
Schemas:
    Customer:
  	x-entity:
    	  key: [id]
  	properties:
    	  name:
          type: string
    	  email:
          type: string
    	  1:
          type: string
    	  ...
paths:
  "/customers/{customerId}":
    get:
  	x-entityProvider:
    	  keyMapping:
          customerId: id
  	parameters:
    	  - in: query
          name: summaryView
          default: true
          schema:
          type: boolean
  	...
  	responses:
    	'200':
         content:
            application/json:
          	  schema:
            	$ref: "#/components/schemas/Customer"
----

==== RAML Example

----
types:
  Customer:

>     (entity):

Author: ```suggestion
    (dg.entity):
```

      key: [id]
    properties:
    	name:
        type: string
    	email:
        type: string
    	id:
        type: string
    	…

/customer/{customerId}:
  uriParameters:
    customerId: string
  get:

>     (entityProvider):

Author: ```suggestion
    (dg.entityProvider):
```

      keyMapping:
        customerId: id
    responses:
      200:
        application/json:
          body: Customer
----

== DataGraph Extensions

> These extensions define behaviors specific for DataGraph. 

Reviewer 2: The same as entity, all extensions are in the same group.

* Operations
* Any type
* Object properties

=== Structure

The target is not specified because it is implicit in the operation.

[%header%autowidth.spread]
|===
|Field |Type |Description
| `datagraph` |Map |A container for DataGraph-specific annotations.
| `datagraph.hide` |Boolean |If `true` the element is hidden in the unified schema. If `false` the element is not hidden, which provides the same result as not setting the annotation.
|===

=== Customization

Each annotation should result in a `HideCustomization`.

=== OAS Example

----
components:
  Schemas:
    Customer:
  	x-datagraph:
    	  hide: true
  	properties:
    	  name:
          type: string
    	  email:
          type: string
  	    x-datagraph:
    	      hide: true
    	  id:
          type: string
    	  ...
paths:
  "/customers/{customerId}":
    get:
  	x-datagraph:
    	  hide: true
  	parameters:
    	  - in: query
          name: summaryView
          default: true
          schema:
          type: boolean
  	...
  	responses:
    	'200':
         content:
            application/json:
          	  schema:
            	$ref: "#/components/schemas/Customer"
----

=== RAML Example

----
types:
  Customer:

>     (datagraph):

Author: ```suggestion
    (dg.datagraph):
```

      hide: true
    properties:
    	name:
        type: string
    	email:
        type: string

>         (datagraph):

Author: ```suggestion
        (dg.datagraph):
```

          hide: true
    	id:
        type: string
    	…

/customer/{customerId}:
  uriParameters:
    customerId: string
  get:

>     (datagraph):

Author: ```suggestion
    (dg.datagraph):
```

This extension applies to:

* Operations
* Any type
* Object properties

==== Structure

The target is not specified because it is implicit in the operation.

[%header%autowidth.spread]
|===
|Field |Type |Description
| `datagraph` |Map |A container for DataGraph-specific annotations.
| `datagraph.name` |String |The desired name for the element in DataGraph. If possible, use a pattern for valid names.
|===

==== Customization

Each annotation should result in a `RenameCustomization`.

==== OAS Example

----
components:
  Schemas:
    Customer:
  	x-datagraph:
    	  name: SalesCustomer
  	properties:
    	  name:
          type: string
    	  e-mail:
          type: string
  	    x-datagraph:
    	      name: email
    	  id:
          type: string
    	  ...
paths:
  "/customers/{customerId}":
    get:
  	x-datagraph:
    	  name: getCustomer
  	parameters:
    	  - in: query
          name: summaryView
          default: true
          schema:
          type: boolean
  	...
  	responses:
    	'200':
         content:
            application/json:
          	  schema:
            	$ref: "#/components/schemas/Customer"
----

==== RAML Example

----
types:
  Customer:

>     (datagraph):

Author: ```suggestion
    (dg.datagraph):
```

      name: SalesCustomer
    properties:
    	name:
        type: string
    	e-mail:
        type: string

>           (datagraph):

Author: ```suggestion
          (dg.datagraph):
```

            name: email
    	id:
        type: string
    	…

/customer/{customerId}:
  uriParameters:
    customerId: string
  get:

>     (datagraph):

Author: ```suggestion
    (dg.datagraph):
```
